src/compiler/crystal/semantic/main_visitor.cr

```diff
@@ -1246,12 +1246,53 @@ module Crystal
       end
 
       # Check if it's ->LibFoo.foo, so we deduce the type from that method
-      if node.args.empty? && obj && (obj_type = obj.type).is_a?(LibType)
-        matching_fun = obj_type.lookup_first_def(node.name, false)
-        node.raise "undefined fun '#{node.name}' for #{obj_type}" unless matching_fun
+      if obj.type.is_a?(LibType)
+        matching_fun = obj.type.lookup_first_def(node.name, false).as(External?)
+        node.raise "undefined fun '#{node.name}' for #{obj.type}" unless matching_fun
 
-        call.args = matching_fun.args.map_with_index do |arg, i|
-          Var.new("arg#{i}", arg.type).as(ASTNode)
+        if node.args.empty?
+          call.args = matching_fun.args.map_with_index do |arg, i|
+            Var.new("arg#{i}", arg.type).as(ASTNode)
+          end
+        else
+          # Variadic funs are always expanded according to input types, due to
+          # different ABIs.
+          if matching_fun.varargs?
+            expand(node)
+            return false
+          end
+
+          # If it's something like `->LibFoo.foo(Bar)`, check that the supplied
+          # parameter types are compatible with the C fun.
+          # This is partially based on `Call#check_fun_arg_type_matches`
+          unless node.args.size == matching_fun.args.size
+            call.wrong_number_of_arguments "'#{call.full_name(obj.type)}'", node.args.size, matching_fun.args.size
+          end
+
+          if node.args.size == matching_fun.args.size
```

Comment:
*suggestion:* This condition seems unnecessary because the previous conditional already checks for the inverse and raises.